### PR TITLE
Add: Type Interface for methods returning Country, State , City inste…

### PR DESCRIPTION
**Pull request by harpreetkhalsagtbit**
issue: #14

### CODE REVIEW

**Comment by harpreetkhalsagtbit** [comment]
> shouldn't you re-export the interfaces in the export {} so it's available for import by consumers?

I mot getting what you are suggesting...

**Comment by baywet** [comment]
Sorry I wasn't clear in my explanation
Currently the interfaces are defined in a file called interfaces.
Although they are referenced and used in the index.ts, which is the main entry point of the package, they are not exported in the module "tree".
If as a consumer I want to use one of the interfaces, let say I have a country in my component state, I'm not sure `import { ICountry } from 'country-state-city';` will work as it currently is defined. And the reason for that is because interfaces are not re-exported in the main module.
Now if you've already tested this scenario and it works, you can disregard my comment, however if you haven't, I suggested to test (and fix) for that as well.

I hope it does make more sense now :) thanks for the follow up and the continuous effort!

**Comment by harpreetkhalsagtbit** [comment]
export interfaces in the main `index.ts` file?

**Comment by harpreetkhalsagtbit** [comment]
import countryList from './lib/country.json';
import stateList from './lib/state.json';
import cityList from './lib/city.json';
import { ICountry, ICity, IState} from './src/interface';

 export = {
	getCountryById: function (id: string): ICountry {
		return _findEntry(countryList, id);
	},
	getStateById: function (id: string): IState {
		return _findEntry(stateList, id);
	},
	getCityById: function (id: string): ICity {
		return _findEntry(cityList, id);
	},
	getStatesOfCountry: function (countryId: string): IState[] {
		var states = stateList.filter(function (value, index) {
			return value.country_id === countryId
		})
		return states.sort(compare)
	},
	getCitiesOfState: function (stateId: string): ICity[] {
		var cities = cityList.filter(function (value, index) {
			return value.state_id === stateId
		})
		return cities.sort(compare)
	},
	getAllCountries: function (): ICountry[] {
		return countryList;
	},
	ICountry,
	ICity,
	IState
}

**Comment by baywet** [comment]
if this is the right syntax to do it, yes! Or something looking like that at least.

**Comment by harpreetkhalsagtbit** [comment]
Not working... 
I will export in typescript way as in standard ES6 using `export default`. Although user with `require()` have to use it as 

`var csc = require('country-state-city').default`

I will mention it in README.md as a breaking chnge, not compatible with previous version

**Comment by baywet** [comment]
I'd suggest to bump the major version of the package at the same time then. This way people with ^ won't "automatically get a breaking change".

**Comment by harpreetkhalsagtbit** [comment]
yes... sure

**Comment by harpreetkhalsagtbit** [comment]
@baywet 
Ready for release, please review again.

1. export = {} changed to export default
2. Interface type re-exported from index.ts
3. Tested with ES6 module syntax
4. Tested with AMD Module - using `require('../index').default`
5. Add tests for Interface Re-Exports
6. Add separate test cases for AMD Modules and ES6 modules

Great learning so far.

File: README.md [hunk 1]
```diff
@@ -9,14 +9,37 @@ https://github.com/hiiamrohit/Countries-States-Cities-database
...
+  - ES6 Module usage
+   
+     `import csc from 'country-state-city'`
+
+  - AMD Module usage
+  
+    `let csc = require('country-state-city').default`
+
+
+## For versions `v0.1.8 and below`
```
Comment (baywet) [inline README.md]: Small typo bellow

Comment (harpreetkhalsagtbit) [reply README.md]: not a type :)

**Comment by web-app-merchants** [comment]
Screenshot from 2021-07-04 09-14-03

@baywet @harpreetkhalsagtbit , got this error message when I updated to new version.

When I use vs code auto import, it works but gives this.
```ts
import type { ICountry } from 'country-state-city/dist/lib/interface';
```

**Comment by honia19** [comment]
> Screenshot from 2021-07-04 09-14-03
> 
> @baywet @harpreetkhalsagtbit , got this error message when I updated to new version.
> 
> When I use vs code auto import, it works but gives this.
> 
> ```ts
> import type { ICountry } from 'country-state-city/dist/lib/interface';
> ```

Yes, the same issue. It doesn't work how in docs `import type { ICountry } from 'country-state-city'`

**Comment by harpreetkhalsagtbit** [comment]
@honia19 Sorry, I missed it during last release. But I am unable to export them correctly now. Can you raise a PR for this? @baywet @honia19

**Comment by honia19** [comment]
> @honia19 Sorry, I missed it during last release. But I am unable to export them correctly now. Can you raise a PR for this? @baywet @honia19 

Yes, okay

**Comment by honia19** [comment]
> @honia19 Sorry, I missed it during last release. But I am unable to export them correctly now. Can you raise a PR for this? @baywet @honia19

What branch do I need to use for PR? Master or develop? @harpreetkhalsagtbit

**Comment by harpreetkhalsagtbit** [comment]
@honia19 develop

**Comment by honia19** [comment]
@harpreetkhalsagtbit Could you provide me access for pushing my branch in origin that creates a pull request with fixed export types

**Comment by harpreetkhalsagtbit** [comment]
@honia19 Please fork the repo and send a PR from that fork.

**Comment by honia19** [comment]
> @honia19 Please fork the repo and send a PR from that fork.

@harpreetkhalsagtbit done, [PR](https://github.com/harpreetkhalsagtbit/country-state-city/pull/100) was opened

**Comment by developerpagadi** [comment]
[!] Error: 'default' is not exported by node_modules/country-state-city/lib/assets/country.json, imported by node_modules/country-state-city/lib/country.js

**Comment by developerpagadi** [comment]
> [!] Error: 'default' is not exported by node_modules/country-state-city/lib/assets/country.json, imported by node_modules/country-state-city/lib/country.js

import countryList from './assets/country.json';
import { findEntryByCode } from './utils';
// Get a country by isoCode.
function getCountryByCode(isoCode) {
    if (!isoCode)
        return undefined;
    return findEntryByCode(countryList, isoCode);
}
// Get a list of all countries.
function getAllCountries() {
    return countryList;
}
export default {
    getCountryByCode,
    getAllCountries,
};

**Comment by developerpagadi** [comment]
> > [!] Error: 'default' is not exported by node_modules/country-state-city/lib/assets/country.json, imported by node_modules/country-state-city/lib/country.js
> 
> import countryList from './assets/country.json'; import { findEntryByCode } from './utils'; // Get a country by isoCode. function getCountryByCode(isoCode) { if (!isoCode) return undefined; return findEntryByCode(countryList, isoCode); } // Get a list of all countries. function getAllCountries() { return countryList; } export default { getCountryByCode, getAllCountries, };

getting error

**Comment by developerpagadi** [comment]
> > > [!] Error: 'default' is not exported by node_modules/country-state-city/lib/assets/country.json, imported by node_modules/country-state-city/lib/country.js
> > 
> > 
> > import countryList from './assets/country.json'; import { findEntryByCode } from './utils'; // Get a country by isoCode. function getCountryByCode(isoCode) { if (!isoCode) return undefined; return findEntryByCode(countryList, isoCode); } // Get a list of all countries. function getAllCountries() { return countryList; } export default { getCountryByCode, getAllCountries, };
> 
> getting error

Error: 'default' is not exported by node_modules/country-state-city/lib/assets/country.json, imported by node_modules/country-state-city/lib/country.js
https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module
node_modules/country-state-city/lib/country.js (1:7)
1: import countryList from './assets/country.json';
          ^
2: import { findEntryByCode } from './utils';
3: // Get a country by isoCode.
Error: 'default' is not exported by node_modules/country-state-city/lib/assets/country.json, imported by node_modules/country-state-city/lib/country.js
    at error (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:7917:30)
    at Module.error (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:9859:16)
    at Module.traceVariable (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:10240:29)
    at ModuleScope.findVariable (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:9018:39)
    at FunctionScope.findVariable (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:3240:38)
    at ChildScope.findVariable (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:3240:38)
    at Identifier.bind (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:4169:40)
    at CallExpression.bind (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:2486:31)
    at CallExpression.bind (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:6589:15)
    at ReturnStatement.bind (/home/akhilkumarpagidi/sites/user-management-v2/node_modules/rollup/dist/shared/rollup.js:2490:23)

**Pull request merged**